internal/scheduler/legacyscheduler.go

```diff
@@ -95,595 +97,579 @@ func SchedulingConstraintsFromSchedulingConfig(
+func NewRescheduler(
+	constraints SchedulingConstraints,
+	config configuration.SchedulingConfig,
+	jobRepo JobRepository,
+	nodeDb *NodeDb,
+	priorityFactorByQueue map[string]float64,
+	initialUsageByQueueAndPriority map[string]schedulerobjects.QuantityByPriorityAndResourceType,
+	schedulingReportsRepository *SchedulingReportsRepository,
+) *Rescheduler {
+	return &Rescheduler{
+		constraints:                           constraints,
+		config:                                config,
+		nodePreemptibleEvictionProbability:    config.Preemption.NodeEvictionProbability,
+		nodeOversubscribedEvictionProbability: config.Preemption.NodeOversubscriptionEvictionProbability,
+		jobRepo:                               jobRepo,
+		nodeDb:                                nodeDb,
+		priorityFactorByQueue:                 maps.Clone(priorityFactorByQueue),
+		usageByQueueAndPriority:               armadamaps.DeepCopy(initialUsageByQueueAndPriority),
+		schedulingReportsRepository:           schedulingReportsRepository,
 	}
 }
 
-func (it *QueuedGangIterator) Next() ([]LegacySchedulerJob, error) {
-	if v, err := it.Peek(); err != nil {
-		return nil, err
-	} else {
-		if err := it.Clear(); err != nil {
-			return nil, err
-		}
-		return v, nil
-	}
+// SchedulerResult is returned by Rescheduler.Schedule().
+type SchedulerResult struct {
+	// Running jobs that should be preempted.
+	PreemptedJobs []LegacySchedulerJob
+	// Queued jobs that should be scheduled.
+	ScheduledJobs []LegacySchedulerJob
+	// For each preempted job, maps the job id to the node on which the job is running.
+	// For each scheduled job, maps the job id to the node on which the job should be scheduled.
+	NodeByJobId map[string]*schedulerobjects.Node
+	// Resource usage by queue, accounting for preempted and scheduled jobs.
+	UsageByQueueAndPriority map[string]schedulerobjects.QuantityByPriorityAndResourceType
 }
 
-func (it *QueuedGangIterator) Clear() error {
-	it.next = nil
-	return nil
-}
+// Reschedule
+// - preempts jobs belonging to queues with total allocation above their fair share and
+// - schedules new jobs belonging to queues with total allocation less than their fair share.
+func (sch *Rescheduler) Schedule(ctx context.Context) (*SchedulerResult, error) {
+	log := ctxlogrus.Extract(ctx)
+	log = log.WithField("function", "Reschedule")
+	preemptedJobsById := make(map[string]LegacySchedulerJob)
+	scheduledJobsById := make(map[string]LegacySchedulerJob)
+	log.Infof(
+		"starting rescheduling with total resources %s",
+		sch.constraints.TotalResources.CompactString(),
+	)
 
-func (it *QueuedGangIterator) Peek() ([]LegacySchedulerJob, error) {
-	if it.hitLookbackLimit() {
-		return nil, nil
-	}
+	// NodeDb snapshot prior to making any changes.
+	// We compare against this snapshot after scheduling to detect changes.
+	snapshot := sch.nodeDb.Txn(false)
 
-	if it.next != nil {
-		return it.next, nil
+	// Evict preemptible jobs.
+	evictorResult, inMemoryJobRepo, err := sch.evict(
```

Comment:
the following ~20  lines are duplicated again just below- we could probably extract this to it's own function.

Reply:
Note that the next eviction uses a different evictor.